### PR TITLE
Implement Customer Import

### PR DESCRIPTION
**Pull request by x4d3**

### CODE REVIEW

File: api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb [hunk 1]
```diff
@@ -117,6 +121,78 @@ def invite_member
       @user = user.confirmed? ? invite : user
     end
 
+    def download_batch_example
+      path = File.join(File.dirname(File.expand_path(__FILE__)), '../../../../../assets/batch-example.csv')
```
Comment (x4d3) [inline api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb]: Sorry, did not find a better way to make a reference to a ressource inside the gem.

File: api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb [hunk 2]
```diff
@@ -160,5 +236,36 @@ def update_app_list
...
+        email = row['email']
+        if email.present? && Devise.email_regexp
+          errors << "Row: #{index}, Invalid email: ''#{email}''" unless email =~ Devise.email_regexp
+        end
+        password = row['password']
+        if password.present?
+          if Devise.password_regex && password !~ Devise.password_regex
+            password << "Row: #{index}, Invalid password'"
+          end
+          errors << "Row: #{index}, Invalid password, It's too short, minimum length is 6." if password.length < 6
```
Comment (x4d3) [inline api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb]: Default value of MnoHUb

File: api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb [hunk 3]
```diff
@@ -117,6 +122,78 @@ def invite_member
...
+
+    # POST /mnoe/jpi/v1/admin/organization/batch_import
+    # Organizations and Users batch import endpoint
+    # The file will be entirely rejected if one line has a validation failure
+    # Each line can be used to create a new customer organisation or a new customer user.
+    # Please use an external reference (external_id) to identify a company if you wish to add multiple users to the same company
+    # the import tool will not identify organisation based on names and will create a new organisation if no external reference is used.
+    # When an external reference if found, the organisation record is updated.
+    # If user email address already exists, it will be linked to the customer
+    def batch_import
```
Comment (BrunoChauvet) [inline api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb]: All this logic should be moved into another class (eg: `ImportCsvOrganization`)

`Controller`
```ruby
def batch_import
  # Check mandatory param(s)
  # Call CSV importer
  @results = ImportCsvOrganization.process(params[:file])
  render 'batch_import'
rescue =>e
  render_bad_request('batch_import', e.message)
end
```

`ImportCsvOrganization`
```ruby
def self.process(file_content)
  # validate and raise error
  # process lines
end
```

File: api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb [hunk 4]
```diff
@@ -117,6 +122,78 @@ def invite_member
...
+      errors = validate_csv(csv)
+      if errors.any?
+        return render json: errors, status: :bad_request
+      end
+      @added_organizations = []
+      @updated_organizations = []
+      @added_users = []
+      @updated_users = []
+      # TODO Move to a Job
+      csv.each do |row|
```
Comment (BrunoChauvet) [inline api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb]: `each_with_index do |row, index|`

File: api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb [hunk 5]
```diff
@@ -117,6 +122,78 @@ def invite_member
...
+        user = MnoEnterprise::User.where(email: row['email']).first
+        if user
+          @updated_users << user
+        else
+          # User created by batch upload don't need confirmation
+          user = MnoEnterprise::User.new(confirmed_at: Time.now)
+          @added_users << user
+        end
+        # Created users don't need any kinf of confirmation
+        user.skip_reconfirmation!
```
Comment (BrunoChauvet) [inline api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb]: Should we skip this only for newly created users?

File: api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb [hunk 6]
```diff
@@ -117,6 +122,78 @@ def invite_member
...
+        user.skip_confirmation_notification!
+        user.name = row['name']
+        user.surname = row['surname']
+        user.phone = row['phone']
+        user.email = row['email']
+        user.password = row['password'].presence || Settings&.admin_panel&.customer_batch_import&.default_password || 'Password1234'
```
Comment (BrunoChauvet) [inline api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb]: Hardcoded password :imp:

File: api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb [hunk 7]
```diff
@@ -160,5 +237,36 @@ def update_app_list
         end
       end
     end
+
+    # TODO: Move to OrganizationCsvValidator
+    def validate_csv(csv)
+      errors = []
+      missing_headers = REQUIRED_HEADERS - csv.headers
+      if (missing_headers.any?)
+        errors.<< "Headers are missing: #{missing_headers.to_sentence}"
```
Comment (BrunoChauvet) [inline api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb]: `errors.<<`

File: api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb (hunk 2 above)
Comment (BrunoChauvet) [reply api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb]: Can you put this magic number in a constant?

File: api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb (hunk 3 above)
Comment (BrunoChauvet) [reply api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb]: Then you can easily unit test different scenarios

File: api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb [hunk 8]
```diff
@@ -117,6 +122,78 @@ def invite_member
       @user = user.confirmed? ? invite : user
     end
 
+    def download_batch_example
+      path = File.join(File.dirname(File.expand_path(__FILE__)), '../../../../../assets/batch-example.csv')
```
[the remark this answers is not in the document]
Comment (x4d3) [reply api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb]: I'll try. But I'm afraid it won't be working as it's  going to point to the express rails app, and not the gem.

File: api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb (hunk 3 above)
Comment (x4d3) [reply api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb]: Yes. I wish to still be able to return the errors as an array, as it's what is being displayed on the front-end.

File: api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb (hunk 4 above)
Comment (x4d3) [reply api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb]: Unfortunately CSV does not implement that method.

File: api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb [hunk 9]
```diff
@@ -147,8 +146,22 @@ def unfreeze
...
-    protected
+    def download_batch_example
+      path = File.join(File.dirname(File.expand_path(__FILE__)), '../../../../../assets/batch-example.csv')
+      send_file(path, filename: 'batch-example.csv', type: 'application/csv')
+    end
 
+    # POST /mnoe/jpi/v1/admin/organization/batch_import
+    def batch_import
+      file = params[:file]
+      # get the file's temporal path
```
Comment (BrunoChauvet) [inline api/app/controllers/mno_enterprise/jpi/v1/admin/organizations_controller.rb]: :clock1: `temporary`

File: api/app/views/mno_enterprise/jpi/v1/admin/organizations/batch_import.json.jbuilder [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+      json.partial! 'mno_enterprise/jpi/v1/admin/users/user', user: u
+    end
+  end
+  json.updated do
+    json.array! users[:updated] do |u|
+      json.partial! 'mno_enterprise/jpi/v1/admin/users/user', user: u
+    end
+  end
+end
+
```
Comment (BrunoChauvet) [inline api/app/views/mno_enterprise/jpi/v1/admin/organizations/batch_import.json.jbuilder]: Your `Enter` key got stuck I think

**Pull request merged**